exchange/exchange.go

```diff
@@ -756,11 +756,31 @@ func applyCategoryMapping(ctx context.Context, requestExt *openrtb_ext.ExtReques
 					} else {
 						// An older bid from a different seatBid we've already finished with
 						oldSeatBid := (seatBids)[dupe.bidderName]
+						rejections = updateRejections(rejections, dupe.bidID, "Bid was deduplicated")
 						if len(oldSeatBid.bids) == 1 {
 							seatBidsToRemove = append(seatBidsToRemove, dupe.bidderName)
-							rejections = updateRejections(rejections, dupe.bidID, "Bid was deduplicated")
 						} else {
-							oldSeatBid.bids = append(oldSeatBid.bids[:dupe.bidIndex], oldSeatBid.bids[dupe.bidIndex+1:]...)
+							// This is a very unique, but still possible case where bid needs to be removed from already processed bidder
+							// This happens when current processing bidder has a bid that has same deduplication key as a bid from already processed bidder
+							// and already processed bid was selected to be removed
+							// See example of input data in unit test `TestCategoryMappingTwoBiddersManyBidsEachNoCategorySamePrice`
+							// Need to remove bid by name, not index in this case
```

Comment:
Can we make the entirety of the else branch into its own function?
```
762              } else {
763                  // This is a very unique, but still possible case where bid needs to be removed from already processed bidder
764                  // This happens when current processing bidder has a bid that has same deduplication key as a bid from already processed bidder
765                  // and already processed bid was selected to be removed
766                  // See example of input data in unit test `TestCategoryMappingTwoBiddersManyBidsEachNoCategorySamePrice`
767                  // Need to remove bid by name, not index in this case
    +                removeByName(dupe.bidID, oldSeatBid.bids)
768 -
769 -                //Find index of bid to remove
770 -                dupeBidIndex := -1
771 -                for i, bid := range oldSeatBid.bids {
772 -                    if bid.bid.ID == dupe.bidID {
773 -                        dupeBidIndex = i
774 -                        break
775 -                    }
776 -                }
777 -                if dupeBidIndex != -1 {
778 -                    if dupeBidIndex < len(oldSeatBid.bids)-1 {
779 -                        oldSeatBid.bids = append(oldSeatBid.bids[:dupeBidIndex], oldSeatBid.bids[dupeBidIndex+1:]...)
780 -                    } else if dupeBidIndex == len(oldSeatBid.bids)-1 {
781 -                        oldSeatBid.bids = oldSeatBid.bids[:len(oldSeatBid.bids)-1]
782 -                    }
783 -                }
784              }
785          }
786          delete(res, dupe.bidID)
exchange/exchange.go
```
I find the `TestCategoryMappingTwoBiddersManyBidsEachNoCategorySamePrice` pretty solid, but if you agree to move the else statement to its own function, can we also add a unit test for it? I don't think we should remove the `TestCategoryMappingTwoBiddersManyBidsEachNoCategorySamePrice` just add a test for `removeByName(string, []*pbsOrtbBid)` (if you decide to name it that).

Reply:
Yes, good idea, refactored!